### PR TITLE
feat(backend): A* solver with MST heuristic, test suite, CI, and BE/ layer

### PR DESCRIPTION
**Pull request by jayenliao**
## What this PR delivers

Full backend implementation for the Divine Spider A* maze solver, ready for frontend integration.

### Core algorithm (`BE/backend.py`, `BE/maze.py`)

- **Multi-objective A\*** — state is `(position, frozenset[remaining objectives])`, so revisiting a cell with a different remaining set is treated as a distinct state
- **MST heuristic** — Prim's algorithm over the current position and all remaining objectives; admissible, giving a tight lower bound that keeps the search efficient
- **Generator interface** — `get_Astar_result(maze)` yields each explored `(row, col)` for animation and returns the optimal path via `StopIteration.value`; returns `[]` when no path exists
- **Input validation** — raises `ValueError` on missing start or objectives
- **Visited set** — skips stale heap entries to avoid redundant expansions
- **`Maze` class** — accepts either a file path or a raw maze string; exposes `getStart()`, `getObjectives()`, `getDimensions()`, `getNeighbors()`, `isValidPath()`

For the full API reference and maze file format, see [`BE/README.md`](https://github.com/XDturtleXD/Divine-Spider-Astar/tree/feature/jay-backend/BE).

### Test suite (`BE/tests/test_sample.py`) — 22 tests, all passing

| Class | Coverage |
| --- | --- |
| `TestPathfinding` | Optimal path length matches BFS on single-goal mazes (small, medium, large) |
| `TestMultiGoal` | All objectives visited in a two-goal maze |
| `TestUnreachable` | Returns `[]` when goal is walled off |
| `TestMaze` | Parsing correctness; string/file equivalence; neighbor counts |
| `TestIsValidPath` | Rejects wall traversal, skipped goals, non-consecutive steps, wrong end position |

### CI (`.github/workflows/tests.yml`)

Runs `uv run pytest` on every push and pull request (Python 3.13, ubuntu-latest).

### Project structure

```
BE/          backend layer (maze.py, backend.py, tests/)
FE/          frontend layer (WIP)
main.py  demo script; wires BE end-to-end
```

For setup and QA instructions, see [`BE/README.md`](https://github.com/XDturtleXD/Divine-Spider-Astar/tree/feature/jay-backend/BE).

## Test plan

- [ ] CI passes (all 22 tests green)
- [ ] `uv run python main.py` prints `Valid?  : Valid` for all four test cases
- [ ] Follow the [QA edge-case checklist](https://github.com/XDturtleXD/Divine-Spider-Astar/tree/feature/jay-backend/BE#5-edge-case-checklist)

### CODE REVIEW

File: BE/maze.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+            (row + 1, col), # down
+            (row - 1, col), # up
+            (row, col + 1), # right
+            (row, col - 1), # left
+        ]
+        neighbors: list[Pos] = []
+        for r, c in possibleNeighbors:
+            if self.isValidMove(r, c):
+                neighbors.append((r, c))
+        self.__states_explored += 1
```
Comment (CHINHUICHU) [inline BE/maze.py]: 雖然這可能不太影響功能，但 `getNeighbors` 應該預期是不會 export 給前端用的？如果直接 call 它，states explored counter 會髒掉嗎（？

File: BE/README.md [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+| File | Role |
+| --- | --- |
+| `BE/maze.py` | the `Maze` class that parses a `.txt` maze file and exposes grid info and neighbor queries |
+| `BE/backend.py` | the search algorithm `get_Astar_result(maze)`, which is a **generator** that streams explored positions and returns the final path |
+| `BE/tests/` | pytest suite for the backend |
+| `main.py` | demo script at project root — exercises the backend end-to-end |
+
+## Backend Overview
+
+This module solves multi-objective maze problems using A\* search with an MST-based heuristic. It provides a `Maze` class for parsing maze files and a generator function `get_Astar_result(maze)` that yields explored positions and returns the optimal path. The backend is designed to be used with a frontend that can visualize the search process and the resulting path.
```
Comment (CHINHUICHU) [inline BE/README.md:18]: 感覺這段好像和第一段有點重複，看要不要精簡一下（？
但這很 minor，也可以先這樣，之後我再整理成一份 project readme

File: BE/README.md [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+#### Full usage (explored positions + path)
+
+```python
+explored = []
+path = []
+try:
+    while True:
+        pos = next(gen)    # (row, col) of each node A* expands
+        explored.append(pos)
+except StopIteration as e:
```
Comment (CHINHUICHU) [inline BE/README.md:106]: 我們目前有針對前端給的 input 做 validation 嗎？感覺要回 error & 如果超過 BE 設定的 input size 上限也要處理

File: BE/README.md [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    path = e.value
```
Comment (CHINHUICHU) [inline BE/README.md]: 這邊感覺還要加上 isValidPath? 看來起現在的後面預期的 flow 是
1. 前端建 maze
2. 後端回傳 position
3. stop -> 後端回傳 path
4. 前端 call isValidPath 確定 path 是不是 valid （可能在 render 上需要做不同處理）

File: BE/maze.py (hunk 1 above)
Comment (jayenliao) [reply BE/maze.py]: I separated `states_explored` increment out of `_getNeighbors`.

File: BE/README.md (hunk 3 above)
Comment (jayenliao) [reply BE/README.md:106]: Patched.

File: BE/README.md (hunk 4 above)
Comment (jayenliao) [reply BE/README.md]: Addressed.

File: BE/maze.py [hunk 5]
```diff
@@ -55,13 +70,27 @@ def __init__(self, filename: str) -> None:
...
+        start_count = 0
         for row in range(len(self.mazeRaw)):
             for col in range(len(self.mazeRaw[0])):
                 if self.mazeRaw[row][col] == self.__startChar:
                     self.__start = (row, col)
+                    start_count += 1
                 elif self.mazeRaw[row][col] == self.__objectiveChar:
                     self.__objective.append((row, col))
 
+        if start_count == 0:
```
Comment (CHINHUICHU) [inline BE/maze.py:82]: [minor] 這邊可以針對有 validation 的 case 寫一下測試嗎？不一定需要在這支 PR 處理

**Review by CHINHUICHU — APPROVED** [review]

File: BE/maze.py (hunk 5 above)
Comment (jayenliao) [reply BE/maze.py:82]: @CHINHUICHU 你可能開個 issue，我們先 merge這個PR到main讓前端可以接著做事？
